# NF-FRAGMENTOMICS PIPELINE

## Introduction

Recent research indicates that analyzing the fragmentation patterns of circulating free DNA (cfDNA) from genome sequencing data can provide insights into nucleosome occupancy in the original cells. In accessible genomic regions, nucleosomes are arranged systematically to facilitate access for DNA-binding proteins.

![cfDNA degradation](assets/img/cfDNA_degradation.png)


## Pipeline Summary

This pipeline calculate the composite coverage over specific genome regions.

![pipeline schema](assets/img/nf-fragmentomics_base.png)

BAM input files are filtered by size (cfDNA fragments have size between 90 and 150 bp) and corrected for GC-bias using the method proposed by [Benjamini & Speed (2012). Nucleic Acids Research, 40(10)].

The input bam is then converted in a coverage file (bigWiggle), at this stage, it is possible to use a blackList of genomic regions. I am blacklisting the Problematic Regions of the Genome defined by ENCODE. [Amemiya, H.M., Kundaje, A. & Boyle, A.P. The ENCODE Blacklist: Identification of Problematic Regions of the Genome. Sci Rep 9, 9354 (2019). https://doi.org/10.1038/s41598-019-45839-z]

With [computeMatrix](https://deeptools.readthedocs.io/en/latest/content/tools/computeMatrix.html) we calculate the coverage of the signal (bam file) on a set of targets (bed file). In my analysis this is typically a set of Transcription Factor Binding Sites regions. We are using the `reference-point` mode: we compute the signal distribution relative to a point (the TFBS center point in our case), expanding on both sides for **4kb**.

A first plot of the composite coverage on the TFBS target set (raw coverage) is generated by the process [plotHeatmap](https://deeptools.readthedocs.io/en/latest/content/tools/plotHeatmap.html).

![plotHeatmap example](assets/img/plotHeatmap_example.png)

The process `PEAK_STATS` generate 3 output files:

 - `peak_data.tsv` file contains the composite coverage (`raw`) for each bin (`bin`),the coverage relative to the background median (`relative`), the background median (`background_median`) for each bin.

 - `peak_stats.csv` summary statistics for the composite coverage: signal, target, source, integration (Monte Carlo integration), length of the peak (`length`), relative length of the peak (`rlength`), mim and max raw values (`ymin` and `ymax`) and the ratio between length and integration (`ratio`)

 - `PeakIntegration.pdf` pdf plot of the calculated metrics:

![plotStats plot](assets/img/peakStats.png)

## Quick Start

<!-- TODO -->

## Documentation

<!-- TODO -->

## Credits

<!-- TODO -->

## Contributions and Support

<!-- TODO -->

## Citation

<!-- TODO -->